src/deepsparse/server/openai_server.py

```diff
@@ -0,0 +1,385 @@
+                index=output.index,
+                message=ChatMessage(role="assistant", content=output.text),
+                finish_reason=output.finish_reason,
+            )
+            choices.append(choice_data)
+
+        num_prompt_tokens = len(final_res.prompt_token_ids)
+        num_generated_tokens = sum(
+            len(output.token_ids) for output in final_res.outputs
+        )
+        usage = UsageInfo(
+            prompt_tokens=num_prompt_tokens,
+            completion_tokens=num_generated_tokens,
+            total_tokens=num_prompt_tokens + num_generated_tokens,
+        )
+        response = ChatCompletionResponse(
+            id=request_id,
+            created=created_time,
+            model=proxy_pipeline.pipeline.model_path,
+            choices=choices,
+            usage=usage,
+        )
+        return response
+
+
+def map_generation_schema(generation_kwargs: Dict) -> Dict:
+    """
+    Map the ChatCompletionRequest to the TextGenerationInput.
+    :param generation_kwargs input fields given as part of the ChatCompletionRequest
+    :returns: updated generated_kwargs, mapped to the TextGenerationInput while
+    raising errors for any properties which are not yet supported.
+    """
+    for k in list(generation_kwargs.keys()):
+        if k in OPENAI_CHAT_NOT_SUPPORTED:
+            return create_error_response(
+                HTTPStatus.BAD_REQUEST, f"{k} is not currently supported"
+            )
+        if k in OPENAI_TO_DEEPSPARSE_MAPPINGS:
+            generation_kwargs[OPENAI_TO_DEEPSPARSE_MAPPINGS[k]] = generation_kwargs[k]
+
+    if generation_kwargs["num_return_sequences"] > 1:
+        generation_kwargs["do_sample"] = True
+
+    return generation_kwargs
+
+
+def create_stream_response_json(
+    index: int,
+    text: str,
+    request_id: str,
+    created_time: int,
+    pipeline: Pipeline,
+    finish_reason: Optional[str] = None,
+) -> str:
+    choice_data = ChatCompletionResponseStreamChoice(
+        index=index,
+        delta=DeltaMessage(content=text),
+        finish_reason=finish_reason,
+    )
+    response = ChatCompletionStreamResponse(
+        id=request_id,
+        created=created_time,
+        model=pipeline.model_path,
+        choices=[choice_data],
+    )
+    response_json = response.json(ensure_ascii=False)
+
+    return response_json
+
+
+async def completion_stream_generator(
+    request, result_generator, request_id, created_time, pipeline
+) -> AsyncGenerator[str, None]:
+    # First chunk with role
+    for i in range(request.n):
+        choice_data = ChatCompletionResponseStreamChoice(
+            index=i,
+            delta=DeltaMessage(role="assistant"),
+            finish_reason=None,
+        )
+        chunk = ChatCompletionStreamResponse(
+            id=request_id, choices=[choice_data], model=pipeline.model_path
+        )
+        data = chunk.json(exclude_unset=True, ensure_ascii=False)
+        yield f"data: {data}\n\n"
```

Comment:
what is the intended usage/consumption for this `data: ...` format?

Reply:
for streaming, just used to separate each streamed token. not necessary

Reply:
needed for the openai api. streaming wont work without it